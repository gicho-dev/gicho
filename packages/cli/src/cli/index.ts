#!/usr/bin/env node

// TODO: WIP...

console.log('[WIP] CLI is not ready yet.')
