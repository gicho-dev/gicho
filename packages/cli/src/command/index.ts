export * from './center'
export * from './error'
export * from './parser'
export * from './types'
