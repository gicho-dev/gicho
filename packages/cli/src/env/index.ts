/**
 * Determines whether the current environment is a CI server.
 */
export const isCI = (): boolean => process.env.CI === 'true'

/**
 * Determines whether the current environment is Windows.
 */
export const isWindows = process.platform.startsWith('win')
