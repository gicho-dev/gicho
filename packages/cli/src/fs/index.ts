export * from './clean-project'
export * from './find-file'
export * from './fs'
export * from './glob'
