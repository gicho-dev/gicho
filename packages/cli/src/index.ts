export * from './env'
export * from './fs'
export * from './terminal'
export * from './utils'
