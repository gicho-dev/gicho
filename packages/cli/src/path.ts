import {
	basename,
	dirname,
	extname,
	format,
	isAbsolute,
	join,
	matchesGlob,
	normalize,
	parse,
	relative,
	resolve,
	toNamespacedPath,
} from 'node:path/posix'

export {
	basename,
	dirname,
	extname,
	format,
	isAbsolute,
	join,
	matchesGlob,
	normalize,
	parse,
	relative,
	resolve,
	toNamespacedPath,
}
