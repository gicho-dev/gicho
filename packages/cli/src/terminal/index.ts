import { createTerminal } from './terminal'

export * from './ansi'
export * from './prompts'
export * from './terminal'
export * from './types'

export const term = createTerminal()
