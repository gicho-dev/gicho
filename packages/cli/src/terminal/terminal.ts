import type {
	LogMessage,
	Terminal,
	TerminalAnsi,
	TerminalConfig,
	TerminalCoreLogType,
	TerminalLogConfig,
	TerminalOptions,
	TerminalPlugin,
	TerminalPrompt,
	TerminalPromptConfig,
} from './types'

import { WriteStream } from 'node:tty'
import { formatWithOptions } from 'node:util'
import { mergeConfigs, mergeConfigsInto } from '@gicho/core/object'

import { ansiCursor, ansiErase, ansiScroll, ansiStyle } from './ansi'
import { getCurrentStackTrace, SYMBOLS, unicodeOr } from './terminal.internal'

/* ----------------------------------------
 *   Terminal Ansi
 * ------------------------------------- */

const terminalAnsiMethods: TerminalPlugin<TerminalAnsi> = (ctx) => {
	return {
		/* ---------- Cursor ---------- */

		cursorTo: (x, y) => ctx.write(ansiCursor.to(x, y)),
		cursorMove: (dx, dy) => ctx.write(ansiCursor.move(dx, dy)),

		cursorUp: (n = 1) => ctx.write(ansiCursor.up(n)),
		cursorDown: (n = 1) => ctx.write(ansiCursor.down(n)),
		cursorLeft: (n = 1) => ctx.write(ansiCursor.left(n)),
		cursorRight: (n = 1) => ctx.write(ansiCursor.right(n)),
		cursorLineStart: () => ctx.write(ansiCursor.lineStart),
		cursorLineEnd: () => ctx.write(ansiCursor.lineEnd),
		cursorNextLine: (n = 1) => ctx.write(ansiCursor.nextLine(n)),
		cursorPrevLine: (n = 1) => ctx.write(ansiCursor.prevLine(n)),
		cursorHide: () => ctx.write(ansiCursor.hide),
		cursorShow: () => ctx.write(ansiCursor.show),
		cursorSave: () => ctx.write(ansiCursor.save),
		cursorRestore: () => ctx.write(ansiCursor.restore),

		/* ---------- Scroll ---------- */

		scrollUp: (n = 1) => ctx.write(ansiScroll.up(n)),
		scrollDown: (n = 1) => ctx.write(ansiScroll.down(n)),

		/* ---------- Clear ---------- */

		clearLine: (dir = 0) =>
			ctx.write(dir === -1 ? ansiErase.lineStart : dir === 1 ? ansiErase.lineEnd : ansiErase.line),
		clearLines: (n = 1) => (n < 1 ? false : ctx.write(ansiErase.lines(n))),

		clearScreen: () => ctx.write(ansiErase.screen),
		clearScreenUp: () => ctx.write(ansiErase.up),
		clearScreenDown: () => ctx.write(ansiErase.down),

		resetScreen: () => ctx.write(ansiErase.reset),
	}
}

/* ----------------------------------------
 *   Terminal Log
 * ------------------------------------- */

const defaultTerminalLogConfig: TerminalLogConfig = {
	formatter(data) {
		const messages = data.messages || []
		const message = messages[0]
		const args = messages.slice(1)

		const prefix = data.prefix ? `${data.prefix} ` : ''
		const suffix = data.suffix ?? ''
		const content = prefix + formatWithOptions({ colors: true }, message, ...args) + suffix

		return content
	},

	level: 3,

	types: {
		// Level 0
		fatal: {
			level: 0,
			prefix: ansiStyle.bgRed(` FATAL `),
			suffix: `\n`,
		},
		error: {
			level: 0,
			prefix: ansiStyle.bgRed(` ERROR `),
			suffix: `\n`,
		},
		// Level 1
		warn: {
			level: 1,
			prefix: ansiStyle.bgYellow(' WARN '),
		},

		// Level 2
		log: {
			level: 2,
		},

		// Level 3
		failure: {
			level: 3,
			prefix: unicodeOr('❌', 'x'),
		},
		info: {
			level: 3,
			prefix: unicodeOr('ℹ️ ', 'i'),
		},
		success: {
			level: 3,
			prefix: unicodeOr('✅', 'x'),
		},

		// Level 4
		debug: {
			level: 4,
			prefix: unicodeOr('⚙', 'D'),
		},

		// Level 5
		trace: {
			level: 5,
			prefix: '→',
			formatter: (data, ctx) => {
				const _message = ctx.config.log.formatter(data, ctx)

				const indent = ' '.repeat(2)

				const stackLines =
					indent +
					getCurrentStackTrace()
						.map((line) =>
							line
								.replace(/^at +/, (m) => ansiStyle.gray(m))
								.replace(/\((.+)\)/, (_, m) => `(${ansiStyle.cyan(m)})`),
						)
						.join(`\n${indent}`)

				return `${_message}\n${stackLines}\n`
			},
		},

		// Level `Verbose` (Infinity)
		verbose: {
			level: Number.POSITIVE_INFINITY,
			prefix: ansiStyle.bgGray(' VERBOSE '),
		},
	},
}

/* ----------------------------------------
 *   Terminal Prompt
 * ------------------------------------- */

const defaultTerminalPromptConfig: TerminalPromptConfig = {
	actions: new Set(['cancel', 'ok', 'up', 'down', 'left', 'right', 'space']),

	colors: {
		base: 'gray',

		initial: 'cyan',
		active: 'cyan',
		completed: 'green',
		canceled: 'red',
		error: 'yellow',

		focused: 'green',
	},

	lineGap: 1,

	S: {
		caret: unicodeOr('█', '_'),
		spacer: ' '.repeat(2),

		start: unicodeOr('╭', 'p'), // '⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏', ●
		bar: unicodeOr('┆', '|'),
		end: unicodeOr('╰', 'b'),

		initial: unicodeOr('●', '*'),
		active: unicodeOr('●', '*'),
		completed: unicodeOr('✔', 'v'),
		canceled: unicodeOr('✕', 'x'),
		error: unicodeOr('▲', 'x'),

		checkbox0: unicodeOr('□', '[ ]'), // checkbox
		checkbox1: unicodeOr('□', '[•]'), // checkbox focused
		checkbox2: unicodeOr('■', '[+]'), // checkbox selected
		radio0: unicodeOr('○', ' '), // radio
		radio1: unicodeOr('⦿', '>'), // radio focused(selected) ◉
	},

	shortcutKeys: {
		return: 'enter',
		escape: 'cancel',
		'\x03': 'cancel',
		h: 'left',
		j: 'down',
		k: 'up',
		l: 'right',
		y: 'ok',
	},
}

const terminalPromptMethods: TerminalPlugin<TerminalPrompt> = (ctx, cfg) => {
	const { style, write } = ctx

	return {
		_P: {
			getMatchedAction: (keys) => {
				for (const key of keys) {
					if (typeof key === 'string') {
						const action = cfg.prompt.actions.has(key) ? key : cfg.prompt.shortcutKeys[key]
						if (action) return action
					}
				}
				return false
			},

			line: (colorKey, symbolKey, str, noNewLine) => {
				const { colors, S } = cfg.prompt
				const symbolText = style[colors[colorKey]](S[symbolKey])
				return `${symbolText}${str ? `${S.spacer}${str}` : ''}${noNewLine ? '' : '\n'}`
			},
			lineRaw: (symbol, str, noNewLine) => {
				const { S } = cfg.prompt
				return `${symbol}${str ? `${S.spacer}${str}` : ''}${noNewLine ? '' : '\n'}`
			},

			linePrefix: () => {
				const { colors, lineGap, S } = cfg.prompt
				return `${style[colors.base](S.bar)}\n`.repeat(lineGap)
			},
			lineBar: (colorKey, str, noNewLine) => ctx._P.line(colorKey, 'bar', str, noNewLine),
			lineEnd: (colorKey, str, noNewLine) => ctx._P.line(colorKey, 'end', str, noNewLine),
		},

		isCanceledPrompt: (result: unknown): result is symbol => result === SYMBOLS.PROMPT_CANCELED,

		startPrompt(message = '') {
			write(ctx._P.line('base', 'start', message))
		},

		endPrompt(message = '') {
			write(ctx._P.linePrefix() + ctx._P.lineEnd('base', message))
		},
	}
}

/* ----------------------------------------
 *   Terminal (Core)
 * ------------------------------------- */

const defaultConfig: TerminalConfig = {
	input: process.stdin,
	output: process.stdout,
	plugins: [],

	log: defaultTerminalLogConfig,
	prompt: defaultTerminalPromptConfig,

	__mock: {
		columns: 80,
		rows: 24,
	},
}

/**
 * Create a terminal object with plugins
 */
export function createTerminal<Ps extends readonly TerminalPlugin<any, any>[]>(
	options: TerminalOptions<Ps> = {},
): Terminal<Ps> {
	const config = mergeConfigs(defaultConfig, options) as TerminalConfig
	const { input, output, plugins, log: logCfg } = config

	const ctx = {
		config,
		input,
		output,
		style: ansiStyle,

		get columns() {
			return output instanceof WriteStream ? output.columns : config.__mock.columns
		},
		get rows() {
			return output instanceof WriteStream ? output.rows : 0
		},

		updateConfig: (options) => {
			mergeConfigsInto(config, options)
		},

		write: (str) => output.write(str),
		writeArgs: (separator = '', ...args) => output.write(args.join(separator)),
	} as Terminal<[]>

	Object.assign(ctx, terminalAnsiMethods(ctx, config))
	Object.assign(ctx, terminalPromptMethods(ctx, config))

	const writeLog: Terminal<Ps>['writeLog'] = (type, messages, options) => {
		const logTypeConfig = mergeConfigs(options, logCfg.types[type])
		const { formatter = logCfg.formatter, level, prefix, suffix } = logTypeConfig
		const logData: LogMessage = {
			messages,
			level,
			prefix,
			suffix,
			type,
		}

		if (level > logCfg.level) return

		const message = formatter(logData, { baseFormatter: logCfg.formatter, config })
		ctx.write(`${message}\n`)
	}

	for (const type in logCfg.types) {
		ctx[type as TerminalCoreLogType] = (...args) => writeLog(type, args)
	}

	for (const plugin of plugins) Object.assign(ctx, plugin(ctx, config))

	return ctx as Terminal<Ps>
}
