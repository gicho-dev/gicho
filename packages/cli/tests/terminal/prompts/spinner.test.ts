import type { Terminal, TerminalPlugin, TerminalSpinner } from '../../../src/terminal'

import EventEmitter from 'node:events'
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test, vi } from 'vitest'

import { createTerminal, promptSpinner } from '../../../src/terminal'
import { MockWritable } from '../terminal.test-utils'

describe.each(['true', 'false'])('terminal/prompts/spinner (isCI = %s)', (isCI) => {
	let origCI: string | undefined
	let output: MockWritable

	let term: Terminal<[TerminalPlugin<TerminalSpinner>]>

	beforeAll(() => {
		origCI = process.env.CI
		process.env.CI = isCI
	})
	afterAll(() => {
		process.env.CI = origCI
	})

	beforeEach(() => {
		output = new MockWritable()
		term = createTerminal({ plugins: [promptSpinner()], output })
		vi.useFakeTimers()
	})
	afterEach(() => {
		vi.useRealTimers()
		vi.restoreAllMocks()
	})

	describe('basic', () => {
		test('renders frames at interval', () => {
			const result = term.spinner()
			result.start()
			for (let i = 0; i < 4; i++) vi.advanceTimersByTime(80) // 4 frames
			result.stop()

			expect(output.buffer).toMatchSnapshot()
		})

		test('renders timer when indicator is "timer"', () => {
			const result = term.spinner({ indicator: 'timer' })
			result.start('START')
			vi.advanceTimersByTime(80)
			result.stop('STOP')

			expect(output.buffer).toMatchSnapshot()
		})

		test('renders cancel symbol if code = 1', () => {
			const result = term.spinner()
			result.start()
			vi.advanceTimersByTime(80)
			result.stop('STOP', 1)
			vi.advanceTimersByTime(80)

			expect(output.buffer).toMatchSnapshot()
		})

		test('renders error message if code > 1', () => {
			const result = term.spinner()
			result.start()
			vi.advanceTimersByTime(80)
			result.stop('STOP', 2)
			vi.advanceTimersByTime(80)

			expect(output.buffer).toMatchSnapshot()
		})

		test('renders messages (with multiple lines)', () => {
			const result = term.spinner()
			result.start('START')
			vi.advanceTimersByTime(80)
			result.message('message 1')
			vi.advanceTimersByTime(80)
			result.message('Long Text\n'.repeat(5).slice(0, -1)) // need prevLine * (5- 1)
			vi.advanceTimersByTime(80)
			result.message('Long Text\n'.repeat(10).slice(0, -1)) // need prevLine * (10- 1)
			vi.advanceTimersByTime(80)
			result.message('message 2')
			vi.advanceTimersByTime(80)
			result.stop('STOP')
			vi.advanceTimersByTime(80)

			expect(output.buffer).toMatchSnapshot()
		})
	})

	describe('custom', () => {
		test('custom symbols and symbolColor', () => {
			const result = term.spinner({ symbolColor: 'red', symbols: ['🍎', '🍌', '🍇', '🍓'] })
			result.start()
			for (let i = 0; i < 4; i++) vi.advanceTimersByTime(80) // 4 frames
			result.stop()

			expect(output.buffer).toMatchSnapshot()
		})

		test('custom delay', () => {
			const result = term.spinner({ delay: 200 })
			result.start()
			for (let i = 0; i < 4; i++) vi.advanceTimersByTime(200) // 4 frames
			result.stop()

			expect(output.buffer).toMatchSnapshot()
		})
	})

	describe('process exit handling', () => {
		let emitter: EventEmitter

		beforeEach(() => {
			emitter = new EventEmitter()

			vi.spyOn(process, 'on').mockImplementation((e, listener) => {
				emitter.on(e, listener)
				return process
			})
			vi.spyOn(process, 'removeListener').mockImplementation((e, listener) => {
				emitter.removeListener(e, listener)
				return process
			})
		})

		afterEach(() => {
			emitter.removeAllListeners()
		})

		test('uses custom cancelMessage when SIGINT', () => {
			const result = term.spinner({ cancelMessage: 'Wow! Canceled!' })
			result.start('Test')
			emitter.emit('SIGINT')

			expect(output.buffer).toMatchSnapshot()
		})

		test('uses custom errorMessage when exit code 2', () => {
			const result = term.spinner({ errorMessage: 'Wow! Error!' })
			result.start('Test')
			emitter.emit('exit', 2)

			expect(output.buffer).toMatchSnapshot()
		})
	})

	test('can be aborted by signal', () => {
		const controller = new AbortController()
		const result = term.spinner({ signal: controller.signal })
		result.start('Signal Test')
		controller.abort()

		expect(output.buffer).toMatchSnapshot()
	})
})
