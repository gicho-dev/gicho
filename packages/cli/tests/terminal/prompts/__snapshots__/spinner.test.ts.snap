// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`terminal/prompts/spinner (isCI = false) > basic > renders cancel symbol if code = 1 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{31red}✕{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > basic > renders error message if code > 1 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{31red}▲{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > basic > renders frames at interval 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◐{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◓{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◑{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > basic > renders messages (with multiple lines) 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  START",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◐{39x}  message 1",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◓{39x}  Long Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text",
  "{CURSOR:prevLine * 4}",
  "{ERASE:down}",
  "{36cyan}◑{39x}  Long Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text",
  "{CURSOR:prevLine * 9}",
  "{ERASE:down}",
  "{36cyan}◒{39x}  message 2.",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{32green}✔{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > basic > renders timer when indicator is "timer" 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  START [0s]",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{32green}✔{39x}  STOP [0s]\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > can be aborted by signal 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}✕{39x}  Canceled\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > custom > custom delay 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◐{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◓{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{36cyan}◑{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > custom > custom symbols and symbolColor 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}🍎{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{31red}🍌{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{31red}🍇{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{31red}🍓{39x}",
  "{CURSOR:x * 1}",
  "{ERASE:down}",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > process exit handling > uses custom cancelMessage when SIGINT 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}✕{39x}  Wow! Canceled!\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = false) > process exit handling > uses custom errorMessage when exit code 2 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}▲{39x}  Wow! Error!\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > basic > renders cancel symbol if code = 1 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  ...",
  "\\n",
  "{31red}✕{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > basic > renders error message if code > 1 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  ...",
  "\\n",
  "{31red}▲{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > basic > renders frames at interval 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  ...",
  "\\n",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > basic > renders messages (with multiple lines) 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  START...",
  "\\n",
  "{36cyan}◐{39x}  message 1...",
  "\\n",
  "{36cyan}◓{39x}  Long Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text...",
  "\\n",
  "{36cyan}◑{39x}  Long Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text\\nLong Text...",
  "\\n",
  "{36cyan}◒{39x}  message 2...",
  "\\n",
  "{32green}✔{39x}  STOP\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > basic > renders timer when indicator is "timer" 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  START...",
  "\\n",
  "{32green}✔{39x}  STOP [0s]\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > can be aborted by signal 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}✕{39x}  Canceled\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > custom > custom delay 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{36cyan}◒{39x}  ...",
  "\\n",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > custom > custom symbols and symbolColor 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}🍎{39x}  ...",
  "\\n",
  "{32green}✔{39x}\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > process exit handling > uses custom cancelMessage when SIGINT 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}✕{39x}  Wow! Canceled!\\n",
  "{CURSOR:show}",
]
`;

exports[`terminal/prompts/spinner (isCI = true) > process exit handling > uses custom errorMessage when exit code 2 1`] = `
[
  "{CURSOR:hide}",
  "{90brightBlack}┆{39x}\\n",
  "{31red}▲{39x}  Wow! Error!\\n",
  "{CURSOR:show}",
]
`;
